783 - draft applications

frankduncan opened the pull request:
Add html and csv of applications in draft status.  To test, look under the reports section, and then use the download this report link to see csv.

@cecilia-donnelly please review as with the other reports, there will probably be some iterating on what exactly does

@jasonaowen I've added to seed.sql (for integration tests, as well as just making development go more smoothly), and am unsure the proper way we want to go about this, so please take a look.

@PaulMorris any comments and suggestions are appreciated, but I've tried to emulate what you talked about in #800

PaulMorris commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/report/ApplicationsInDraftReportController.java:
```diff
@@ -0,0 +1,159 @@
+    @RequestMapping(value = "/admin/reports/draftapplications.csv", method = RequestMethod.GET)
+    public void getEnrollments(HttpServletResponse response) throws PortalServiceException {
+        String csvFileName = "draftapplications" +
+            LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd-hhmm")) +
```
Would it be worth abstracting out a function for formatting the date stamp, and then calling it for each of the reports?  That way they are always consistently formatted.  And/or, I guess that pattern string could just be defined somewhere and then referenced/used for each report.

frankduncan replied:
I was definitely on the edge of this.  I'll create a ReportControllerUtils.  There will probably be other overlaps coming as we do more reports.

PaulMorris commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/report/ApplicationsInDraftReportController.java:
```diff
@@ -0,0 +1,159 @@
+            CSVPrinter csvPrinter = new CSVPrinter(response.getWriter(), CSVFormat.DEFAULT);
+
+            csvPrinter.printRecord(
+                "Month with Draft",
```
"Month in Draft" ?

PaulMorris commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/report/ApplicationsInDraftReportController.java:
```diff
@@ -0,0 +1,159 @@
+                enrollmentMonths.add(em);
+
+                for (Enrollment enrollment : enrollments) {
+                    em.addEnrollment(enrollment);
```
I'm probably missing something, but won't this end up adding all of the `enrollments` to every month in `enrollmentMonths`?

PaulMorris commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/report/ApplicationsInDraftReportController.java:
```diff
@@ -0,0 +1,159 @@
+                    enrollment.getSubmissionDate() == null ||
+                    nextMonth.minusDays(1).isBefore(toLocalDate(enrollment.getSubmissionDate()))
+                )
+            ) {
```
Ah, I see, `addEnrollment` doesn't always add an enrollment, it depends on the enrollment's dates.  It might be worth renaming or something to make this clearer?

PaulMorris commented on psm-app/frontend/src/main/js/admin/report.js:
```diff
@@ -0,0 +1,12 @@
+  });
+
+  return data;
+}
```
Looks good!

PaulMorris commented on psm-app/integration-tests/src/test/java/gov/medicaid/features/PsmPage.java:
```diff
@@ -67,6 +67,11 @@ public void checkAccessibility() {
+        String headerText = $("#wrapper h1").getText();
+        assertThat(headerText).isNotEqualTo("Error");
+    }
+
```
Nice!

PaulMorris commented on psm-app/integration-tests/src/test/java/gov/medicaid/features/report/steps/ReportStepDefinitions.java:
```diff
@@ -35,4 +37,30 @@ public void i_download_the_applications_by_reviewer_report() {
+    }
+
+    @Then("^I should see some draft applications for '(.*)'$")
+    public void i_should_see_some_draft_applications_for(String month) throws Exception {
```
Good call on parameterizing months here.

PaulMorris commented on psm-app/integration-tests/src/test/java/gov/medicaid/features/report/steps/ReportSteps.java:
```diff
@@ -3,17 +3,39 @@
 
+    @Step
+    public void applicationsInDraftReport() {
+        reportPage.click$(".applicationsInDraftLink");
```
:+1:

PaulMorris commented on psm-app/integration-tests/src/test/java/gov/medicaid/features/report/steps/ReportSteps.java:
```diff
@@ -3,17 +3,39 @@
 
 public class ReportSteps {
     private ReportPage reportPage;
+    private ApplicationsInDraftPage aidPage;
```
I'd rather not abbreviate 'aid' here, for easier readability.  What about `applicationsInDraftPage` or `inDraftPage` instead?

PaulMorris commented on psm-app/integration-tests/src/test/resources/features/report/accessibility.feature:
```diff
@@ -11,3 +11,9 @@ Feature: Accessibility Checks for Report Pages
+  Scenario: Applications in Draft Page
+    Given I am logged in as an admin
+    And I am on the reports page
+    And I am on the applications in draft page
```
I'd just have one 'And', the one for the draft page, and have it call the 'I am on the reports page' code before navigating to the 'in draft' page.  Which is how I've done others like this.

frankduncan commented on psm-app/cms-web/WebContent/WEB-INF/pages/admin/reports/applications_in_draft_page.jsp:
```diff
@@ -0,0 +1,75 @@
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<!DOCTYPE html>
+<html lang="en-US">
+  <c:set var="title" value="Draft Report"/>
```
I completely forgot that you had come up with names, and that for this one we were going with Draft Applications Report.  So I'm changing that everywhere, in classnames and whatnot.

frankduncan commented:
Thanks @PaulMorris !  I believe I've cleaned up with regard to your review.  As I noted in a buried comment, I've changed all the names to Draft Applications to be in line with your original names we talked about a while ago.

PaulMorris commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/report/DraftApplicationsReportController.java:
```diff
@@ -0,0 +1,156 @@
+
+    @RequestMapping(value = "/admin/reports/draftapplications.csv", method = RequestMethod.GET)
+    public void getEnrollments(HttpServletResponse response) throws PortalServiceException {
+        String csvFileName = ReportControllerUtils.buildCsvName("draftapplications");
```
:+1:

PaulMorris commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/report/DraftApplicationsReportController.java:
```diff
@@ -0,0 +1,156 @@
+                enrollmentMonths.add(em);
+
+                for (Enrollment enrollment : enrollments) {
+                    em.addEnrollmentIfInMonth(enrollment);
```
:+1:

PaulMorris commented on psm-app/integration-tests/src/test/java/gov/medicaid/features/report/steps/ReportStepDefinitions.java:
```diff
@@ -35,4 +37,31 @@ public void i_download_the_applications_by_reviewer_report() {
+    @Given("^I am on the draft applications page$")
+    public void i_am_on_the_draft_applications_page() throws Exception {
+        i_am_on_the_reports_page();
+        reportSteps.draftApplicationsReport();
```
In similar cases we've used "navigateTo..." naming so let's call this method 'navigateToDraftApplicationReport' for consistency and to make it clear what it does.

PaulMorris commented on psm-app/integration-tests/src/test/java/gov/medicaid/features/report/steps/ReportSteps.java:
```diff
@@ -3,17 +3,39 @@
 
+    @Step
+    public void draftApplicationsReport() {
+        reportPage.click$(".draftApplicationsLink");
```
'navigateToDraftApplicationReport' as above.

frankduncan commented:
@PaulMorris I think that's set.

PaulMorris commented:
Screenshots while I was at it:

![screenshot-2018-5-16 draft applications](https://user-images.githubusercontent.com/1091693/40127612-c6e8dbe8-58fd-11e8-9096-aa137c89fd5c.png)

I'm seeing some submission dates, but aren't draft applications the ones that haven't been submitted yet, or am I missing something here?

![screenshot-2018-5-16 draft applications 1](https://user-images.githubusercontent.com/1091693/40127645-d9d22c50-58fd-11e8-9cf6-a88bab70e7e9.png)

cecilia-donnelly commented:
@PaulMorris Thanks for the screenshots!  @frankduncan, please merge when you see this -- having the seed data won't hurt anything, and we can always discuss next week if necessary.  No need to block on that.

frankduncan commented:
About to merge, but I wanted to answer your question first @PaulMorris .  The month of August 2017 will have items that were in draft at the end of that month, so anything submitted after August 31, 2017 will show up in the report.  You'll also see ticket id 1001 in 09-2017 and then drop off for October.
